src/registry.rs

```diff
@@ -0,0 +1,243 @@
+//! This allows the user to resolve a URL to an ObjectStore at runtime. Unlike
+//! [`crate::parse_url`], this allows for custom logic to be executed
+//! when a URL is resolved to an ObjectStore. It also serves as a cache for object stores
+//! to avoid repeated creation.
+
+use crate::{parse_url, ObjectStore};
+use std::collections::HashMap;
+use std::sync::{Arc, RwLock};
+use url::Url;
+
+/// [`ObjectStoreRegistry`] maps a URL to an [`ObjectStore`] instance,
+/// and allows users to read from different [`ObjectStore`]
+/// instances. For example the registry might be configured so that
+///
+/// 1. `s3://my_bucket/lineitem/` mapped to the `/lineitem` path on an
+///    AWS S3 object store bound to `my_bucket`
+///
+/// 2. `s3://my_other_bucket/lineitem/` mapped to the (same)
+///    `/lineitem` path on a *different* AWS S3 object store bound to
+///    `my_other_bucket`
+///
+/// In this particular case, the url `s3://my_bucket/lineitem/` will be provided to
+/// [`ObjectStoreRegistry::get_store`] and one of three things will happen:
+///
+/// - If an [`ObjectStore`] has been registered with [`ObjectStoreRegistry::register_store`] with
+///   `s3://my_bucket`, that [`ObjectStore`] will be returned
+///
+/// - If an AWS S3 object store can be ad-hoc discovered by the url `s3://my_bucket/lineitem/`, this
+///   object store will be registered with key `s3://my_bucket` and returned.
+///
+/// - Otherwise `None` will be returned, indicating that no suitable [`ObjectStore`] could
+///   be found
+///
+/// This allows for two different use-cases:
+///
+/// 1. Systems where object store buckets are explicitly created using DDL, can register these
+///    buckets using [`ObjectStoreRegistry::register_store`]
+///
+/// 2. Systems relying on ad-hoc discovery, without corresponding DDL, can create [`ObjectStore`]
+///    lazily by providing a custom implementation of [`ObjectStoreRegistry`]
+pub trait ObjectStoreRegistry: Send + Sync + std::fmt::Debug + 'static {
+    /// If a store with the same key existed before, it is replaced and returned
+    fn register_store(
+        &self,
+        url: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>>;
+
+    /// Get a suitable store for the provided URL. For example:
+    ///
+    /// - URL with scheme `file:///` or no scheme will return the default LocalFS store
+    /// - URL with scheme `s3://bucket/` will return the S3 store
+    /// - URL with scheme `hdfs://hostname:port/` will return the hdfs store
+    ///
+    /// If no [`ObjectStore`] found for the `url`, ad-hoc discovery may be executed depending on
+    /// the `url` and [`ObjectStoreRegistry`] implementation. An [`ObjectStore`] may be lazily
+    /// created and registered.
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>>;
+
+    /// List all registered URLs
+    fn list_urls(&self) -> Vec<String>;
+}
+
+/// The default [`ObjectStoreRegistry`]
```

Comment:
Having a default registry that opens files makes sense for DataFusion but I think is less clearly the right behavior for the generic `object_store` crate.

We should either:
1.  Document the automatic creation of `LocalFileSystem`  (if that is indeed the behavior we want).
2. Remove the automatic creation of `LocalFileSystem` (or at least make it a config flag)
4. Remove `DefaultObjectStoreRegistry` from this PR

Perhaps something like

```suggestion
/// A  default  in memory [`ObjectStoreRegistry`] that maps a prefix to an ObjectStore
```

Reply:
Oops! I think I actually did remove the local filesystem registration. I must have missed some docs. I'll take a pass and make sure the docs reflect reality. My intention was to remove the default registration for local FS

Reply:
@alamb I want to make sure I understand what you're saying. Right now, the default registry does not automatically create a local filesystem. It starts empty. It _does_, however, auto-create an ObjectStore using `parse_url` for _any_ `get_store` call that results in a cache miss. So if you say `get_store("file:///foo/bar")` and there is no store for `file:///foo`, it will create one.

Are you suggesting that we fully remove the `parse_url` behavior from the `DefaultObjectStoreRegistry` to make it as dumb as possible (essentially just a map)?

Reply:
Thinking about this more: I think we should just make `DefaultObjectStoreRegistry` really dumb by default. Least surprises and all. So I'm inclined to remove `get_url_key` from the `DefaultObjectStoreRegistry` and just have it return `None` if nothing matches a URL get_store key.